File: chexit.py
import argparse
import functools
import logging
import ipaddress
import itertools
import json
import re
import sys

parser = argparse.ArgumentParser()
parser.add_argument('-c',  '--config')
parser.add_argument('-o',  '--output', default=sys.stdout)
parser.add_argument('-v',  '--verbose', action='count', default=0)
parser.add_argument('-e',  '--expand', choices=['addr', 'none'], default='addr')
parser.add_argument('-f',  '--formatter', choices=['json', 'csv'], default='json')
group = parser.add_mutually_exclusive_group()
group.add_argument('-u',  '--uuid',    help='uuid1[:uuid2...]')
group.add_argument('-vp', '--v_polid', help='vdom1,polID1[:vdom2,polID2...]')
args = parser.parse_args()

if args.verbose>0:
    FORMAT = '%(levelname)s:%(message)s'
    LEVEL  = logging.DEBUG
else:
    FORMAT = '%(message)s'
    LEVEL  = logging.INFO

if args.output==sys.stdout:
    logging.basicConfig(stream=args.output,   format=FORMAT, level=LEVEL)
else:
    logging.basicConfig(filename=args.output, filemode='w',
                        format=FORMAT,        level=LEVEL)

def is_match (_match):
    return isinstance(_match, re.Match)

def pipe(args, *funcs):
    return functools.reduce(lambda arg, func: func(arg), funcs, args)

def search_by_uuid(state, line):
    entry = re.sub('\n', '|', line)
    start = r"^\s*edit\s\d+"
    end   = r"^\s*next"

    is_start = lambda: is_match(re.match(start, entry))
    is_end   = lambda: is_match(re.match(end, entry))
    is_found = lambda: is_match(re.search(state['keys'], state['search']))

    match (is_start(), is_found(), is_end()):
        case (True, _, _):
            state['search'] = entry
        case (_, True, True):
            state['found'] = ''.join([state['search'], entry])
            logging.debug('Found {}'.format(state['keys']))
        case _:
            state['search'] = ''.join([state['search'], entry])

    return state

def search_by_v_polid(state, line):
    entry    = re.sub('\n', '|', line)
    pol_id = state['keys'].split(',')[1]
    vdom   = state['keys'].split(',')[0]

    in_global   = lambda: is_match(re.match(r"^\s*config global", entry))
    in_vdom     = lambda: is_match(re.match(r"^\s*edit\s" + vdom, entry))
    in_policy   = lambda: is_match(re.match(r"^\s*edit\s{}[^\d]".format(pol_id), entry))
    in_policies = lambda: is_match(re.match(r"^\s*config firewall policy", entry))

    state['flag'] = {
        ''             : 'Waiting VDOM' if in_global()   else '',
        'Waiting VDOM' : 'In VDOM'      if in_vdom()     else state['flag'],
        'In VDOM'      : 'In policies'  if in_policies() else state['flag'],
        'In policies'  : state['flag']
    }[state.get('flag', '')]

    if re.match(r"^\s*edit\s{}[^\d]".format(pol_id), entry) \
         and state['flag']=='In policies':
        state['search'] = entry
    elif re.match(r"^\s*next", entry) \
         and re.search(pol_id, state['search']):
        state['found'] = ''.join([state['search'], entry])
        dbg = 'Found ID {} in VDOM {}'.format(pol_id, vdom)
        logging.debug(dbg)
    elif state['search'] and state['flag']=='In policies':
        state['search'] = ''.join([state['search'], entry])

    return state

def search_addr_grp(state, line):
    entry = re.sub('\n', '|', line)

    is_subnet  = lambda: is_match(re.search(r"set subnet (.*)\|$", state['search']))
    is_addrgrp = lambda: is_match(re.search(r"set member (.*)\|$", state['search']))

    if state["key"] == 'all':
        state['found'] = {'subnet' : 'all'}
        logging.debug("Found subnet {}".format(state['found']))
    elif re.search('edit "{}"'.format(state["key"]), entry):
        state['search'] = entry
        state['flag'] = 'In address group'
    elif re.search(r"\s*next", entry) and state['flag']=="In address group":
        match (is_subnet(), is_addrgrp()):
            case (True, _):
                match_ = re.search(r"set subnet (.*)\|$", state['search'])
                state['found'] = {'subnet' : match_.group(1)}
                logging.debug("{} : {}".format(state['key'], state['found']))
            case (_, True):
                match_ = re.search(r"set member (.*)\|$", state['search'])
                state['found'] = {'member' : match_.group(1)}
                logging.debug("{} : {}".format(state['key'], state['found']))
    else:
        state['search'] = ''.join([state['search'], entry])

    return state

def fill(info, field):
    return info | dict([split_key_val_in_field(field)])

def trim_keys(found):
    return functools.reduce(fill, found.split('|'), {})

def split_field(field):
    match = re.search(r'^(\w+) (.*)$', field)
    key, val = match.groups()

    return (key, val) if key!='id' else (key, int(val))

def split_key_val_in_field(field):
    key, *value = re.split(' ', field)
    cond = not re.match('^(id|name|action|logtraffic|uuid|comments)', field)

    return (key, value) if cond else split_field(field)

def trim_prfx(found):
    trimmer = lambda STR, RGX: re.sub(RGX[0], RGX[1], STR)
    keyset  = [
        (r'^\s+edit\s(?=\w+)', 'id '),
        (r'(?<=\|)\s+set\s',   ''),
        (r'\|\s+next.*$', ''),
        ('"', '')
    ]

    return functools.reduce(trimmer, keyset, found)

def lookup_key(config_name, key, search_by):
    init = {
        'found'  : '',
        'search' : '',
        'keys'   : key,
        'flag'   : ''
    }

    logging.debug('Looking up {}'.format(key))

    with open(config_name) as config:
        return pipe(
            itertools.accumulate(config, search_by(), initial=init),
            lambda srch: itertools.takewhile(lambda o: not o['found'], srch),
            lambda findings: list(findings).pop()['found']
        )

def add_addr_grp_to_search_or_get_subnet(init, addr):
    addrs, subnets = init
    init_srch = {
        'found'  : '',
        'search' : '',
        'key'    : addrs.pop(0),
        'flag'   : ''
    }

    with open(args.config) as config:
        temp = pipe(
            itertools.accumulate(config, search_addr_grp, initial=init_srch),
            lambda srch: itertools.takewhile(lambda o: not o['found'], srch),
            lambda findings: list(findings).pop()['found']
        )

    match temp:
        case {'subnet' : 'all'}:
            subnets.append('all')
        case {'subnet' : subnet}:
            ip_subnet = ipaddress.ip_network(subnet.replace(" ", "/"))
            subnets.append(str(ip_subnet))
        case {'member' : members}:
            [addrs.append(member.strip('"')) for member in members.split(" ")]

    return (addrs, subnets)

def search_till_subnet_is_found(old_addr_queue, old_subnet_list):
    match len(old_addr_queue):
        case 0:
            return old_subnet_list
        case _:
            length = range(len(old_addr_queue))
            init   = (old_addr_queue, old_subnet_list)
            new_addr_queue, new_subnet_list = functools.reduce(
                add_addr_grp_to_search_or_get_subnet,
                length,
                init
            )

            return search_till_subnet_is_found(
                new_addr_queue,
                new_subnet_list
            )

def expand_subnet_from_addr_grp(output):
    match args.expand:
        case 'none':
            logging.debug("No subnet expansion")
            expansion = {}
        case 'addr':
            logging.debug("Subnet expansion")
            expansion = {
                'srcaddr' : search_till_subnet_is_found(output.get('srcaddr'), []),
                'dstaddr' : search_till_subnet_is_found(output.get('dstaddr'), [])
            }

    return output | expansion

def lookup_keys(config_name, _type, key_list, list_sep=':'):
    def pipe_flow(key):
        return pipe(
            lookup_key(config_name, key, search_by),
            trim_prfx,
            trim_keys,
            expand_subnet_from_addr_grp,
        )

    def search_by():
        return {
            'UUID'           : search_by_uuid,
            'VDOM-AND-POLID' : search_by_v_polid
        }[_type]

    keys = key_list.split(list_sep)
    logging.debug('Number of input, {}: {}'.format(len(keys), keys))

    return [pipe_flow(key) for key in keys]

def format_output(entries, formatter, line_sep=';'):
    def dict_to_string(line, item):
        key, value = item

        match key:
            case 'id':
                return str(value)
            case 'name' | 'uuid' | 'action' | 'logtraffic' | 'comments':
                return line_sep.join([line, value])
            case 'srcintf' | 'dstintf' | 'srcaddr' | 'dstaddr' | \
                 'schedule' | 'service':
                joinedvalues = ','.join(value)
                return line_sep.join([line, joinedvalues])

    match formatter:
        case "json":
            return  pipe(
                entries,
                json.dumps,
                logging.info
            )
        case "csv":
            rows   = [functools.reduce(dict_to_string, entry.items(), '')
                      for entry in entries]
            head   = line_sep.join(entries.pop(0).keys())
            output = ['sep=' + line_sep] + [head] + rows

            return pipe(
                '\n'.join(output),
                logging.info
            )

if __name__ == "__main__":

    if sys.hexversion < 50856688:
        run = '.'.join(map(str, sys.version_info[:3]))
        err = 'chexit requires at least Python 3.8.2; you have {}'.format(run)
        raise RuntimeError(err)

    if args.uuid:
        _type = 'UUID'
        keys  = args.uuid
    elif args.v_polid:
        _type = 'VDOM-AND-POLID'
        keys  = args.v_polid

    pipe(
        lookup_keys(args.config, _type, keys),
        lambda output: format_output(output, args.formatter),
    )

File: chexit.go
package main

import (
	"C"
	"bufio"
	"encoding/json"
	"flag"
	"fmt"
	"maps"
	"net"
	"os"
	"reflect"
	"regexp"
	"slices"
	"strconv"
	"strings"
)

const (
	UUID int = iota
	VDOM_AND_POLID
)

type Config struct {
	Expander     string
	Filename     string
	Keys         string
	Formatter    string
	UUID, vPolID string
	Verbose      int
	SearchBy     int
}

type Policy struct {
	ID       int      `json:"id"`
	Name     string   `json:"name"`
	UUID     string   `json:"uuid"`
	Srcintf  []string `json:"srcintf"`
	DstIntf  []string `json:"dstintf"`
	SrcAddr  []string `json:"srcaddr"`
	DstAddr  []string `json:"dstaddr"`
	Service  []string `json:"service"`
	Schedule []string `json:"schedule"`
	Action   string   `json:"action"`
}

type addrGroup struct {
	subnets  map[string]string
	addrs    []string
	filename string
}

const (
	IN_SEARCH int = iota
	WAIT_VDOM
	IN_VDOM
	IN_POLICIES
)

type state struct {
	flag   int
	found  string
	keys   string
	search string
}

const (
	CLOSE int = iota
	INFO
	DEBUG
	ERROR
)

type Message struct {
	title int
	body  string
}

func Logger(log <-chan Message, verbose int) {
	toStr := []string{INFO: "INFO", DEBUG: "DEBUG", ERROR: "ERROR"}

	for msg := range log {
		if msg.title == CLOSE {
			return
		}

		if verbose > 0 {
			fmt.Printf("%s: %s\n", toStr[msg.title], msg.body)
		} else if msg.title == INFO || msg.title == ERROR {
			fmt.Printf("%s\n", msg.body)
		}
	}
}

func main() {
	config := parseArgs()
	log := make(chan Message)
	go Logger(log, config.Verbose)

	if errors := checkArgs(config); errors != "" {
		log <- Message{ERROR, fmt.Sprintf("Invalid arguments: %s\n", errors)}
		log <- Message{}
		os.Exit(1)
	}

	if config.UUID != "" {
		config.SearchBy = UUID
		config.Keys = config.UUID
	} else if config.vPolID != "" {
		config.SearchBy = VDOM_AND_POLID
		config.Keys = config.vPolID
	}

	log <- Message{INFO, Format(LookUpKeys(log, &config), config.Formatter)}
	log <- Message{}
}

func Format(policies []Policy, formatter string) string {
	var formatted string

	switch formatter {
	case "json":
		jsonStr, _ := json.Marshal(policies)
		formatted = string(jsonStr)
	case "csv":
		for _, policy := range policies {
			formatted = fmt.Sprintf("%s%s", formatted, policytoString(policy))
		}
	}

	return formatted
}

func LookUpKeys(log chan<- Message, config *Config) []Policy {
	var policies []Policy

	for _, key := range strings.Split(config.Keys, ":") {
		policies = append(policies, lookupKey(log, key, config))
	}

	return policies
}

func addAddrOrSubnet(log chan<- Message, group *addrGroup) {
	const (
		IN_GRP int = iota
		IN_ADDR
	)

	var key string
	var flag int

	key, group.addrs = group.addrs[0], group.addrs[1:]
	pattern := fmt.Sprintf(`edit "%s"`, key)
	addrRe := regexp.MustCompile(`set (subnet|member) (.*)$`)

	file, err := os.Open(group.filename)
	if err != nil {
		log <- Message{ERROR, err.Error()}
		os.Exit(1)
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)

	if key == "all" {
		group.subnets["all"] = "all"
		log <- Message{DEBUG, fmt.Sprintf("Found subnet all")}
		return
	}

	for scanner.Scan() {
		entry := scanner.Text()
		addr := addrRe.FindStringSubmatch(entry)

		if ok, _ := regexp.MatchString(`config firewall addrgrp`, entry); ok {
			flag = IN_GRP
		} else if ok, _ := regexp.MatchString(pattern, entry); ok && flag == IN_GRP {
			flag = IN_ADDR
		} else if flag == IN_ADDR && addr != nil {
			if addr[1] == "subnet" {
				addSubnet(addr[2], group)
			} else if addr[1] == "member" {
				addAddrs(addr[2], group)
			}
			log <- Message{DEBUG, fmt.Sprintf("%s: %s", key, addr[2])}
			return
		}
	}

	if err := scanner.Err(); err != nil {
		log <- Message{ERROR, err.Error()}
		os.Exit(1)
	}
}

func addAddrs(addr string, group *addrGroup) {
	for _, member := range strings.Split(addr, " ") {
		group.addrs = append(group.addrs, strings.ReplaceAll(member, `"`, ""))
	}
}

func addSubnet(addr string, group *addrGroup) {
	fields := strings.Split(addr, " ")
	subnet := fmt.Sprintf("%s/%s", fields[0], getCIDR(fields[1]))
	group.subnets[subnet] = subnet
}

func checkArgs(config Config) string {
	var errors []string
	conditions := map[string]bool{
		"no filename":                    config.Filename == "",
		"both UUID & VDOM/Policy ID set": config.UUID != "" && config.vPolID != "",
		"UUID or VDOM/Policy ID not set": config.UUID == "" && config.vPolID == "",
	}

	for msg, condition := range conditions {
		if condition {
			errors = append(errors, msg)
		}
	}

	return strings.Join(errors, ", ")
}

func expandSubnets(log chan<- Message, group *addrGroup) []string {
	group.subnets = make(map[string]string)
	for len(group.addrs) != 0 {
		for _ = range len(group.addrs) {
			addAddrOrSubnet(log, group)
		}
	}

	return slices.Collect(maps.Keys((*group).subnets))
}

func getCIDR(mask string) string {
	length, _ := net.IPMask(net.ParseIP(mask).To4()).Size()
	return strconv.Itoa(length)
}

func lookupKey(log chan<- Message, key string, config *Config) Policy {
	policy := &Policy{}
	var searchBy func(chan<- Message, *state, string)
	state := &state{keys: key}

	log <- Message{DEBUG, fmt.Sprintf("Looking up %s", key)}

	switch config.SearchBy {
	case UUID:
		searchBy = searchByUUID
	case VDOM_AND_POLID:
		searchBy = searchByVDOMAndPolID
	}

	file, err := os.Open(config.Filename)
	if err != nil {
		log <- Message{ERROR, err.Error()}
		os.Exit(1)
	}
	defer file.Close()
	scanner := bufio.NewScanner(file)

	for scanner.Scan() {
		line := scanner.Text()
		searchBy(log, state, line)
		if state.found != "" {
			trimPrefix(&state.found)
			jsonBody, _ := json.Marshal(trimKeys(state.found))
			json.Unmarshal(jsonBody, policy)
			break
		}
	}

	if err := scanner.Err(); err != nil {
		log <- Message{ERROR, err.Error()}
		os.Exit(1)
	}

	if config.Expander == "addr" {
		log <- Message{DEBUG, "Subnet expansion"}
		group := &addrGroup{filename: config.Filename, addrs: policy.SrcAddr}
		policy.SrcAddr = expandSubnets(log, group)
		group.addrs = policy.DstAddr
		policy.DstAddr = expandSubnets(log, group)
	} else if config.Expander == "none" {
		log <- Message{DEBUG, "No subnet expansion"}
	}

	return *policy
}

func parseArgs() Config {
	var config Config

	flag.StringVar(&config.Filename, "config", "", "Configuration file name")
	flag.IntVar(&config.Verbose, "verbose", 0, "Verbosity")
	flag.StringVar(&config.Expander, "expand", "addr", "Expand (none, addr)")
	flag.StringVar(&config.Formatter, "formatter", "json", "Output format (json, csv)")
	flag.StringVar(&config.UUID, "uuid", "", "uuid1[:uuid2...]")
	flag.StringVar(&config.vPolID, "v_polid", "", "vdom1,polID1[:vdom2,polID2...]")

	flag.Parse()

	return config
}

func policytoString(policy Policy) string {
	var key, head, rows string
	sep := ";"
	values := reflect.ValueOf(policy)
	typesOf := values.Type()

	for i := 0; i < values.NumField(); i++ {
		key = strings.ToLower(typesOf.Field(i).Name)

		switch key {
		case "id":
			head = fmt.Sprintf("%s", key)
			rows = fmt.Sprintf("%d", values.Field(i).Interface())
		case "name", "uuid", "action", "logtraffic", "comments":
			head = fmt.Sprintf("%s%s%s", head, sep, key)
			rows = fmt.Sprintf("%s%s%s", rows, sep, values.Field(i).Interface())
		case "srcintf", "dstintf", "srcaddr", "dstaddr", "schedule", "service":
			head = fmt.Sprintf("%s%s%s", head, sep, key)
			values := values.Field(i).Interface().([]string)
			rows = fmt.Sprintf("%s%s%s", rows, sep, strings.Join(values, ","))
		}
	}

	return fmt.Sprintf("sep=%s\n%s\n%s", sep, head, rows)
}

func searchByVDOMAndPolID(log chan<- Message, state *state, line string) {
	var entry string = line + "|"

	fields := strings.Split(state.keys, ",")
	vdom, polID := fields[0], fields[1]
	pattern := fmt.Sprintf(`^\s*edit\s%s[^\d]`, polID)
	foundPolicy := strings.Contains(state.search, polID)

	updateFlag(state, vdom, entry)

	if ok, _ := regexp.MatchString(pattern, entry); ok && state.flag == IN_POLICIES {
		state.search = entry
	} else if ok, _ := regexp.MatchString(`^\s*next`, entry); ok && foundPolicy {
		state.found = fmt.Sprintf("%s%s", state.search, entry)
		log <- Message{DEBUG, fmt.Sprintf("Found ID %s in VDOM %s", polID, vdom)}
	} else if state.search != "" && state.flag == IN_POLICIES {
		state.search = fmt.Sprintf("%s%s", state.search, entry)
	}
}

func searchByUUID(log chan<- Message, state *state, line string) {
	var entry string = line + "|"

	s, _ := regexp.MatchString(`^\s*edit\s\d+`, entry)
	f, _ := regexp.MatchString(`^\s*next`, entry)

	if s {
		state.search = entry
	} else if f && strings.Contains(state.search, state.keys) {
		state.found = fmt.Sprintf("%s%s", state.search, entry)
		log <- Message{DEBUG, fmt.Sprintf("Found %s", state.keys)}
	} else {
		state.search = fmt.Sprintf("%s%s", state.search, entry)
	}
}

func splitField(field string) (string, interface{}) {
	var fields []string = strings.Split(field, " ")

	m, _ := regexp.MatchString(`^(id|name|action|logtraffic|uuid|comments)`, field)
	if !m {
		return fields[0], fields[1:]
	} else if fields[0] != "id" {
		return fields[0], fields[1]
	} else {
		val, _ := strconv.Atoi(fields[1])
		return fields[0], val
	}
}

func trimKeys(found string) map[string]interface{} {
	fields := make(map[string]interface{})

	for _, field := range strings.Split(found, "|") {
		key, val := splitField(field)
		fields[key] = val
	}

	return fields
}

func trimPrefix(found *string) {
	var keyset map[string]string = map[string]string{
		`^\s+edit\s`:   "id ",
		`\|\s+set\s`:   "|",
		`\|\s+next.*$`: "",
		`"`:            "",
	}

	for pattern, replacement := range keyset {
		re := regexp.MustCompile(pattern)
		*found = re.ReplaceAllString(*found, replacement)
	}
}

func updateFlag(state *state, vdom, entry string) {
	switch state.flag {
	case IN_SEARCH:
		if ok, _ := regexp.MatchString(`^\s*config global`, entry); ok {
			state.flag = WAIT_VDOM
		}
	case WAIT_VDOM:
		pattern := fmt.Sprintf(`^\s*edit\s%s`, vdom)
		if ok, _ := regexp.MatchString(pattern, entry); ok {
			state.flag = IN_VDOM
		}
	case IN_VDOM:
		if ok, _ := regexp.MatchString(`\s*config firewall policy`, entry); ok {
			state.flag = IN_POLICIES
		}
	case IN_POLICIES:
		state.flag = IN_POLICIES
	}
}
